### PR TITLE
First working implementation of the derotoation package

### PR DESCRIPTION
**Pull request by lauraporta**
### What is this PR about?
This Pull Request introduces the inaugural implementation of the derotation package, tailored specifically for neuroscience research applications in calcium imaging, with a focus on 3-photon imaging. 

### Recommended Review Strategy
Given the substantive nature of this PR, which encompasses around 2,000 lines of code, I recommend the following review strategy to ensure a thorough understanding and efficient use of reviewers' time:

1. **Initial Discussion**: A walk-through to discuss the scope, mechanics, and intended application of the package, complemented by demonstrations using actual data.

2. **Hands-On Trial**: Personally execute a run of the package to gain firsthand experience with its functionality.

3. **Architectural Review**: Engage in a dialogue to dissect and understand the high-level design and architectural decisions.

Please note, as this is an embryonic stage of a research-focused package, fine-grained scrutiny on software engineering practices, though important, may not be the immediate priority. Contextual information is provided below to aid in the review process.

### Core Functionality
This package is designed to execute derotation-by-line via motor feedback for image correction, enhancing the data quality from 3-photon calcium imaging. It utilizes analog signals from the operational motor to unwarp TIFF images acquired during the imaging process. In addition, it is equipped with the capability to generate diagnostic plots for signal validation and create CSV files encapsulating rotation angle data for each frame.

### Rationale for This PR
The primary goal of this PR is to integrate the initial version of the derotation package into the main branch, thereby streamlining access and utilization for end-users.

### Contributions of This PR
This PR introduces the following components to the project:
- Core classes embodying the package's functionality.
- Example scripts demonstrating the use cases.
- A `README` file that provides a preliminary guide to the package.
- Preliminary tests that cover the critical aspects of the package.

### References
The conceptual foundation for this package stems from the work found in [jvoigts' repository](https://github.com/jvoigts/rotating-2p-image-correction), which offers Matlab-based image derotation techniques, and is further informed by the study from [Voigts and Harnett 2019](https://pubmed.ncbi.nlm.nih.gov/31759808/). For additional insights, refer to [this internal presentation](https://github.com/neuroinformatics-unit/documentation/blob/laura-margrie-lab-presentation/outputs/presentations/margrie_lab_meetings/13102023_laura_lab_meeting/margrie-lab-meeting-october.pdf).

### Testing Approach
The package has undergone extensive local testing with datasets pertinent to its intended use cases. Unit tests have been crafted to validate the most crucial methods. There is a scope for adding more tests to ensure broader coverage.

### Impact Assessment
- Is this a breaking change? Not Applicable.
- Does this PR necessitate updates to the documentation? Not Applicable, as current documentation updates are included within this PR.

## Checklist for Reviewers
- [x] Ensure code functions correctly locally.
- [x] Verify that new test cases cover new functionality.
- [x] Check that documentation is updated to reflect any new changes or additions.
- [x] Confirm that code is formatted in accordance with [pre-commit](https://pre-commit.com/) standards.

### CODE REVIEW

**Review by niksirbi — APPROVED** [review]
Hey @lauraporta , here I will provide a high-level summary of my review. 

More detailed comments are sprinkled throughout, but because of your pushes to the PR during the review, many of them show up at the wrong code lines (now we learned not to do that again). If you are unsure what some these comments are referring to, feel free to ask me. 

As you requested, I focused on test-running the pipelines and reviewing the high-level design and architecture.  I haven't checked the scientific accuracy of the results, that would probably take much more work.

I would recommend only addressing my most trivial comments in this PR, then merging to main, and raising the more substantive comments as issues to be addressed in future PRs.

So here are my high-level impressions:

- I got both rotation pipelines to run on some datasets. I tried it on several datasets, and it seems to run (after fixing the path issues I have pointed out). So yayyy 🎉 
- That said, it took me a while to figure out how to run the pipeline. It would be helpful to have one docs page explaining how to setup the pipeline to run from the user's perspective (even if the user is only you, or your future self)
- I think having configs in a specific folder, and requiring the code to be run from the root directory of the package are quite restrictive choices. In one of my comments I've suggested some alternative design ideas, we can also discuss them in person
- The test coverage is quite low, I would try to increase that in next PRs. Perhaps it's wise to do that before massive refactorings.
- The two pipeline classes are very long and hard to parse from start to finish. I think some refactoring will be needed. For example, you can rewrite many of the general purpose methods as static functions and put them in separate utility modules. Then they can be imported and used by both pipelines. I think it's better to rely more on composition rather than on inheritance.
- Regarding the inheritance: it seems that the incremental pipeline uses only a small subset of the methods in the full rotation pipeline. I wonder whether it's worth defining a `BaseRotationPipeline` class from which both inherit. You can also consider other design patterns for pipelines, for example [chain of responsibility](https://refactoring.guru/design-patterns/chain-of-responsibility), though I haven't thought through how it would exactly apply here.
- Many of the methods inside the pipeline don't need to be public, you can convert those to private by prefixing them with `_`.  Sphinx will then exclude them from API docs by default. 

Overall, well done, this is a substantial piece of work! Must have taken you many hours of hard work.

File: derotation/derotate_by_line.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            rotation_just_finished = not is_rotating and (
+                np.absolute(rot_deg_line[i - 1]) > np.absolute(rotation)
+            )
+
+        if is_rotating:
+            if rotation_completed and (line_counter != 0):
+                # when starting a new rotation in the middle of the image
+                rotated_filled_image = (
+                    np.ones_like(image_stack[image_counter]) * min_value_img
+                )  # non sampled pixels are set to the min val of the image
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply derotation/derotate_by_line.py:68]: Thanks for the feedback Niko. Having this dark pixel was very useful in debugging because it helpted me see clearly when and how the derotation of the analysis made sense. 

I have then changed the code and now I am using the $F_0$ val of the luminance distribution, the same that should then be calculated by suite2p. 

What do you think? I am unsure of filling-in with interpolation. Maybe I can leave the choice to the user and let them decide if they want $F_0$ or the interpolation

File: derotation/load_data/custom_data_loaders.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import numpy as np
+from scipy.io import loadmat
+
+
+def read_randomized_stim_table(path_to_randperm):
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply derotation/load_data/custom_data_loaders.py:5]: I am not sure how much care I should give to these very methods because they are specific to the way a single postdoc in a single lab saves the data. The idea is that people should re-write these methods for their own setup.
I think typing and tests should be there to verify if the new implementation other people will write works. Ideally they will be for an abstract interface of data_loaders (yet to be built).

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        for _ in range(number_of_rotations)
+        for num in intervals
+        for _ in range(rotation_len)
+    ]
+    return dummy_signal
+
+
+@pytest.fixture
+def full_rotation(number_of_rotations, rotation_len, intervals, dummy_signal):
+    dummy_signal += np.random.normal(
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply tests/conftest.py:61]: Thanks, I didn't know we are going to say goodbye to `np.random.normal` 🪦

File: derotation/derotate_by_line.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
+import copy
+
+import numpy as np
+import tqdm
+from scipy.ndimage import rotate
+
+
+def rotate_an_image_array_line_by_line(
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply derotation/derotate_by_line.py:8]: I agree, thanks

File: examples/deformation_by_line_of_a_dog.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ax[1].set_title("Rotated image 1")
+ax[2].imshow(img_rotated[1], cmap="gray")
+ax[2].set_title("Rotated image 2")
+ax[3].imshow(img_rotated[2], cmap="gray")
+ax[3].set_title("Rotated image 3")
+
+ax[0].axis("off")
+ax[1].axis("off")
+ax[2].axis("off")
+ax[3].axis("off")
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply examples/deformation_by_line_of_a_dog.py:33]: I will do this refactoring 👍

File: derotation/load_data/custom_data_loaders.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    speed = pseudo_random["stimulus_random"][:, 0]
+
+    return direction, speed
+
+
+def read_rc2_bin(path_aux, chan_names):
+    n_channels = len(chan_names)
+
+    # Read binary data saved by rc2 (int16)
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply derotation/load_data/custom_data_loaders.py:20]: It is the name of another experiment "roller coaster 2". This is the python translation of a matlab method to read matlab files in another project 🙊

File: derotation/load_data/custom_data_loaders.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    data, chan_names = read_rc2_bin(path_to_aux, channel_names)
+
+    data_dict = {chan: data[:, i] for i, chan in enumerate(chan_names)}
+
+    frame_clock = data_dict["scanimage_frameclock"]
+    line_clock = data_dict["scanimage_lineclock"]
+    full_rotation = data_dict["PI_rotON"]
+    rotation_ticks = data_dict["PI_rotticks"]
+
+    return frame_clock, line_clock, full_rotation, rotation_ticks
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply derotation/load_data/custom_data_loaders.py:42]: It is true, this is not optimal. I am considering refactoring it 👍

File: derotation/analysis/full_rotation_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,882 @@
...
+                rolled_starts[i],
+            )
+        ]
+
+        self.rotation_ticks_peaks = np.delete(
+            self.rotation_ticks_peaks,
+            np.where(
+                np.isin(self.rotation_ticks_peaks, inter_roatation_interval)
+            ),
+        )
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply derotation/analysis/full_rotation_pipeline.py:434]: Ok I'll look into it

File: derotation/analysis/full_rotation_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,912 @@
...
+        config_name : str
+            Name of the config file without extension.
+            Either "full_rotation" or "incremental_rotation".
+
+        Returns
+        -------
+        dict
+            Config dictionary.
+        """
+        path_config = "derotation/config/" + config_name + ".yml"
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply derotation/analysis/full_rotation_pipeline.py:80]: This seems like a good idea, I like it. I can work on this after this PR is merged :) I'll also check how datashuttle does it as well

File: derotation/derotate_by_line.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (niksirbi) [reply derotation/derotate_by_line.py:68]: I think F0 is fine for now. You may add an option for interpolation in a later PR.

File: derotation/load_data/custom_data_loaders.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (niksirbi) [reply derotation/load_data/custom_data_loaders.py:5]: I understand that this implementation is specific to one person's data, but I think it's still useful to have at least some minimal docstrings, because they would be of great help when trying to adapt these functions for another project (at least they would help me understand what kind of loader functions do I need to implement in the first place).

File: derotation/analysis/full_rotation_pipeline.py [hunk 10]
```diff
@@ -0,0 +1,987 @@
...
+            The image for which you want to find the offset.
+
+        Returns
+        -------
+        float
+            The offset.
+        """
+        gm = GaussianMixture(n_components=7, random_state=0).fit(
+            img[0].reshape(-1, 1)
+        )
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply derotation/analysis/full_rotation_pipeline.py:852]: For now, I am following the guidelines of a tutorial of 2p data analysis. I think they have found empirically that 7 in general is a good number of terms for this fit 🤔
